lib/std/hash/xxhash.zig

```diff
@@ -21,20 +17,174 @@ pub const XxHash64 = struct {
     const prime_4 = 0x85EBCA77C2B2AE63; // 0b1000010111101011110010100111011111000010101100101010111001100011
     const prime_5 = 0x27D4EB2F165667C5; // 0b0010011111010100111010110010111100010110010101100110011111000101
 
+    const Accumulator = struct {
+        acc1: u64,
+        acc2: u64,
+        acc3: u64,
+        acc4: u64,
+
+        fn init(seed: u64) Accumulator {
+            return .{
+                .acc1 = seed +% prime_1 +% prime_2,
+                .acc2 = seed +% prime_2,
+                .acc3 = seed,
+                .acc4 = seed -% prime_1,
+            };
+        }
+
+        inline fn updateEmpty(self: *Accumulator, input: anytype, comptime unroll_count: usize) usize {
+            var i: usize = 0;
+
+            if (unroll_count > 0) {
+                const unrolled_bytes = unroll_count * 32;
+                while (i + unrolled_bytes <= input.len) : (i += unrolled_bytes) {
+                    inline for (0..unroll_count) |j| {
+                        self.processStripe(input[i + j * 32 ..][0..32]);
+                    }
+                }
+            }
+
+            while (i + 32 <= input.len) : (i += 32) {
+                self.processStripe(input[i..][0..32]);
+            }
+
+            return i;
+        }
+
+        inline fn processStripe(self: *Accumulator, buf: *const [32]u8) void {
+            self.acc1 = round(self.acc1, mem.readIntLittle(u64, buf[0..8]));
+            self.acc2 = round(self.acc2, mem.readIntLittle(u64, buf[8..16]));
+            self.acc3 = round(self.acc3, mem.readIntLittle(u64, buf[16..24]));
+            self.acc4 = round(self.acc4, mem.readIntLittle(u64, buf[24..32]));
+        }
+
+        inline fn merge(self: Accumulator) u64 {
+            var acc = rotl(u64, self.acc1, 1) +% rotl(u64, self.acc2, 7) +%
+                rotl(u64, self.acc3, 12) +% rotl(u64, self.acc4, 18);
+            acc = mergeAccumulator(acc, self.acc1);
+            acc = mergeAccumulator(acc, self.acc2);
+            acc = mergeAccumulator(acc, self.acc3);
+            acc = mergeAccumulator(acc, self.acc4);
+            return acc;
+        }
+
+        inline fn mergeAccumulator(acc: u64, other: u64) u64 {
+            const a = acc ^ round(0, other);
+            const b = a *% prime_1;
+            return b +% prime_4;
+        }
+    };
+
+    inline fn finalize(
+        unfinished: u64,
+        byte_count: usize,
+        partial: anytype,
+    ) u64 {
+        std.debug.assert(partial.len < 32);
+        var acc = unfinished +% @as(u64, byte_count) +% @as(u64, partial.len);
+
+        switch (partial.len) {
+            inline 0, 1, 2, 3 => |count| {
+                inline for (0..count) |i| acc = finalize1(acc, partial[i]);
+                return avalanche(acc);
+            },
+            inline 4, 5, 6, 7 => |count| {
+                acc = finalize4(acc, partial[0..4]);
+                inline for (4..count + 4) |i| acc = finalize1(acc, partial[i]);
```

Comment:
`count` instead of `count + 4`?

Reply:
Good catch, thanks - perhaps there should be a test for every all partial lengths 0..<32

Reply:
If you look at smhasher you can use the `Sanity` command with an appropriate check value (should be one for the reference xxhash32) which will check all lengths from 1 to 256. I found this useful for wyhash where there are a lot of different paths dependent on length.

e.g.

```
./SMHasher --test=Sanity zig-wyhash
```

Reply:
I've (since you pointed out the bug) ported the sanity tests from the reference repo - just not sure if I should include them. It's ~8322 cases for each function (all lengths from 0 to 4160 with two seeds).

Reply:
I would suggest using the Verification test which can be implemented fairly compactly: https://github.com/tiehuis/smhasher/blob/513a0a6d24f721eab793e574218418cd69a89d9c/KeysetTest.cpp#L16

Reply:
The SMHasher verification test looks like a good idea, but I think I would rather leave porting it to a separate PR - it should probably be used on all the hash functions and I would think we should verify using other implementations (probably via you're SMHasher fork) that we are producing the correct result before we commit the verification results to the repo.